engineering_tools/carma

```diff
@@ -50,54 +50,85 @@ __pull_newest_carma_base() {
-# The method will return the first tag that appears in the config for the possible organizations and images.
-# If the config is not set or an image cannot be found the method will return a carma-base image as the default.
-# The user can specify optional organizations or images as sed compatable regex lists.
-# For example calling 
-#  $(__get_image_from_config 'carma-platform:\|carma-messenger:' 'usdotfhwastol\|usdotfhwastoldev\|usdotfhwastolcandidate')
-# 
+# Method returns a fully specified docker image name and tag based on the currently set carma-config
+# The method extracts DOCKER_ORG and DOCKER_TAG from the .env file in carma-config to look for
+# combinations from SUPPORTED_IMAGES that exist locally.
+# If the config is not set or an image cannot be found the method will return most recent
+# carma-base image as the default.
+# If none is possible from above, returns error message.
+# The user can specify optional images as sed compatable regex lists.
+# For example calling
+#  $(__get_image_from_config 'carma-platform\|carma-messenger')
+#
 #  If no parameters are provided the defaults will be used
 ##
 __get_image_from_config() {
-    local TARGET_IMAGE="usdotfhwastol/carma-base:latest"
-    local SUPPORTED_ORANIZATIONS="usdotfhwastol\|usdotfhwastoldev\|usdotfhwastolcandidate"
-    local SUPPORTED_IMAGES="carma-platform:\|carma-messenger:"
-
-    if [ -n "$1" ]
-    then
-        local SUPPORTED_IMAGES="$1"
-    fi
+    # Default image is the most recent carma-base image
+    local DEFAULT_IMAGE=$(docker images --format '{{.Repository}}:{{.Tag}} {{.CreatedAt}}' | grep -E 'carma-base' | grep -v '<none>' | sort -k 2 -r | head -n 1 | awk '{print $1}')
+    local SUPPORTED_IMAGES="carma-platform\|carma-messenger"
 
-    if [ -n "$2" ]
-    then
-        
-        local SUPPORTED_ORANIZATIONS="$2"
+    if [ -n "$1" ]; then
+        SUPPORTED_IMAGES="$1"
     fi
 
     # Check if carma-config is set. If so, use what was in the carma config
     if docker container inspect carma-config > /dev/null 2>&1; then
-
-        local IMAGE_BACKUP=$TARGET_IMAGE
-
-        # sed command finds the first instance of an image which is part of the set defined in SUPPORTED_ORGANIZATIONS and SUPPORTED_IMAGES.
-        # then returns the name of the image.  
-        local TARGET_IMAGE=$(docker run --rm --volumes-from carma-config:ro --entrypoint sh busybox:latest -c \
-            "sed -n -e '/image:\s*\($SUPPORTED_ORANIZATIONS\)\/\($SUPPORTED_IMAGES\)/ s/.*\image: *//p' /opt/carma/vehicle/config/docker-compose.yml | grep -m 1 '.*'")
-            
-
-        # If the sed resulted in an empty string then reset the tag
-        if [ -z "$TARGET_IMAGE" ]; then
-            local TARGET_IMAGE=$IMAGE_BACKUP
-            echoerr "Could not identify usable image from config so defaulting to image: $TARGET_IMAGE"
+        local IMAGE_FOUND=0
+        # Extract the .env file content from the carma-config volume
```

Comment:
Already mentioned in your comment, but just adding a note, this will use the local path to the .env file if specified

Reply:
Although we discussed this offline, I had a second thought where I realized that we don't need to use local .env for this function. This function `__get_image_from_config` is used in only 3 places:
- carma_exec: which can reliably use carma-config's `.env` most of the time since that would be same behavior as when we didnt use `.env`. if they want to use specific image, `-i` option is available. Making it use local `.env` seemed to be was not straightforward and over-engineering. 
- carma config edit: doesn't have to be any specific image version anyways.
- carma install: we shouldn't use local `.env` since the function is intending to install version in the config. 

So I decided to only keep config's .env here